Add Hycon Technology HY4245 fuel gauge  driver

ssekar15 opened the pull request:
Add a support for Hycon Technology HY4245 fuel gauge driver

ssekar15 committed:
dts: fuel-gauge: hy4245: Add HY4245 fuel gauge

Add dts bindings for Hycon Technology HY4245 fuel gauge.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

DBS06 commented on drivers/fuel_gauge/hy4245/hy4245.c:
```diff
@@ -0,0 +1,160 @@
+
+#define HY4245_CHIPID				0x4245
+
+#define HY4245_CMD_CTRL				0x00
+#define HY4245_CMD_TEMPERATURE			0x06
+#define HY4245_CMD_VOLTAGE			0x08
+#define HY4245_CMD_CURRENT			0x0c
+#define HY4245_CMD_CAPACITY_REM			0x10
+#define HY4245_CMD_CAPACITY_FULL		0x12
+#define HY4245_CMD_AVG_CURRENT			0x14
+#define HY4245_CMD_TIME_TO_EMPTY		0x16
+#define HY4245_CMD_TIME_TO_FULL			0x18
+#define HY4245_CMD_CHRG_VOLTAGE			0x30
+#define HY4245_CMD_CHRG_CURRENT			0x32
+#define HY4245_CMD_CAPACITY_FULL_AVAIL		0x78
+#define HY4245_CMD_RELATIVE_STATE_OF_CHRG	0x2c
```
Maybe convert that to an enum

DBS06 reviewed (changes requested):
I did a quick review and so far it looks good to me.
I am doing a deep review when I am back from my vacation (1st of September)
 
You need to squash your commits and provide more detailed information about the changes in the commit message, as suggested in the guidelines at https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-body

Consider adding an emulation for this IC and add a test, it's not a deal breaker, but it helps us to increase the code coverage and prevents accidentally breaking changes.

Furthermore, add your driver to `tests/drivers/build_all/fuel_gauge/i2c.dtsi`, so it will be at least build during the CI.

ssekar15 committed:
drivers: fuelgauge: Add Hycon Technology HY4245 driver

Add a support for Hycon Technology HY4245 fuel gauge driver.

The HY4245 operates with Single and Two Cells Li+ battery cells as a
stand–along battery gauge. The device uses GaugePackTM algorithm, which
mixes Coulomb–Counting and Open–Circuit–Voltage (OCV) measurements with
battery cell characteristics to manage battery gauge, to maintain accurate
battery capacity estimates with compensation for rate, temperature, age
and self–discharge effects.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
tests: drivers: build_all: fuel_gauge: Add hy4245 device to build test

Add build tests for the hycon,hy4245 device.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

aaronemassey reviewed (dismissed):
Generally LGTM. I agree the DTS change should probably be squashed if only because it doesn't mean much outside, but it's not something I feel strongly about or blocking the PR over.

ssekar15 commented:
@teburd @aaronemassey thanks for review. Re-based to solve conflict, please review

DBS06 reviewed (approved):
I did my review a couple of time ago but unfortunately I forgot to submit it.  
I added two additional commits, but they are not a showstopper.  
Nevertheless, it looks good to me and ready to merge.

msalau reviewed (changes requested):
Few nit-picks from me.

aaronemassey reviewed (dismissed):
LGTM but should consider what [msalau](https://github.com/msalau) mentioned before merging.

ssekar15 commented:
@cfriedt Could you remove the DNM

aaronemassey reviewed (approved):
LGTM. As DBS06 mentioned we can't fully validate because no access to the datasheet but that's not a huge concern in my opinion.

Ideally we'd have just enough emulation for unit tests but again I understand with the NDA that may not be possible here.
